src/graph/transform/knn.cc

```diff
@@ -0,0 +1,117 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file graph/transform/knn.cc
+ * \brief k-nearest-neighbor (KNN) implementation
+ */
+
+#include "kdtree_ndarray_adaptor.h"
+#include "knn.h"
+#include <vector>
+#include <dgl/runtime/registry.h>
+#include <dgl/runtime/packed_func.h>
+#include "../../array/check.h"
+
+using namespace dgl::runtime;
+using namespace dgl::transform::knn_utils;
+namespace dgl {
+namespace transform {
+namespace impl {
+
+/*! \brief The kd-tree implementation of K-Nearest Neighbors */
+template <typename FloatType, typename IdType>
+void KdTreeKNN(const NDArray data_points, const IdArray data_offsets,
+               const NDArray query_points, const IdArray query_offsets,
+               const int k, IdArray result) {
+  int64_t batch_size = data_offsets->shape[0] - 1;
+  int64_t feature_size = data_points->shape[1];
+  IdType * data_offsets_data = static_cast<IdType*>(data_offsets->data);
+  IdType * query_offsets_data = static_cast<IdType*>(query_offsets->data);
+  IdType * result_data = static_cast<IdType*>(result->data);
+  FloatType * query_points_data = static_cast<FloatType*>(query_points->data);
```

Comment:
Missing `const` qualifiers (except `result_data`).  In general, we would prepend all variables that shouldn't be mutated with `const` qualifiers.  Please check for other places as well.

We also no longer use `static_cast` in new code.  Please use `query_offsets.Ptr<IdType>()` to do the same thing.